server/src/main/java/org/elasticsearch/search/diversification/mmr/MMRResultDiversification.java

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.search.diversification.mmr;
+
+import org.apache.lucene.index.VectorSimilarityFunction;
+import org.apache.lucene.search.ScoreDoc;
+import org.elasticsearch.index.mapper.vectors.DenseVectorFieldMapper;
+import org.elasticsearch.search.diversification.ResultDiversification;
+import org.elasticsearch.search.diversification.ResultDiversificationContext;
+import org.elasticsearch.search.rank.RankDoc;
+import org.elasticsearch.search.vectors.VectorData;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class MMRResultDiversification extends ResultDiversification {
+
+    @Override
+    public RankDoc[] diversify(RankDoc[] docs, ResultDiversificationContext diversificationContext) throws IOException {
+        if (docs == null || docs.length == 0 || ((diversificationContext instanceof MMRResultDiversificationContext) == false)) {
+            return docs;
+        }
+
+        MMRResultDiversificationContext context = (MMRResultDiversificationContext) diversificationContext;
+
+        Map<Integer, Integer> docIdIndexMapping = new HashMap<>();
+        for (int i = 0; i < docs.length; i++) {
+            docIdIndexMapping.put(docs[i].doc, i);
+        }
+
+        VectorSimilarityFunction similarityFunction = DenseVectorFieldMapper.VectorSimilarity.MAX_INNER_PRODUCT.vectorSimilarityFunction(
+            context.getIndexVersion(),
+            diversificationContext.getElementType()
+        );
+
+        // our chosen DocIDs to keep
+        List<Integer> selectedDocIds = new ArrayList<>();
+
+        // always add the highest scoring doc to the list
+        int highestScoreDocId = -1;
+        float highestScore = Float.MIN_VALUE;
+        for (ScoreDoc doc : docs) {
+            if (doc.score > highestScore) {
+                highestScoreDocId = doc.doc;
+                highestScore = doc.score;
+            }
+        }
+        selectedDocIds.add(highestScoreDocId);
+
+        // test the vector to see if we are using floats or bytes
+        VectorData firstVec = context.getFieldVector(highestScoreDocId);
+        boolean useFloat = firstVec.isFloat();
+
+        // cache the similarity scores for the query vector vs. searchHits
+        Map<Integer, Float> querySimilarity = getQuerySimilarityForDocs(docs, similarityFunction, useFloat, context);
```

Comment:
So here, we could just do the check on whether it's a float and do a call to just get bytes or floats respectively and the API would be a bit cleaner I think